### PR TITLE
Add the ability to discard history more flexibly

### PR DESCRIPTION
**Pull request by hanselabreu**
Currently the only way to discard history is via the `historic` tag. One problem with this approach is that sometimes you do not have control over the update being sent to the editor, thus making it impossible to append any `tags` to the change.

My proposal is to allow a `discardHistory` option to be passed to `registerHistory` which allows a more dynamic way to prevent history from being pushed.

Example below:
https://github.com/facebook/lexical/assets/27902567/1cacdc9a-0dec-4244-80c5-9f2c30361e4e

### CODE REVIEW

**Comment by etrepum** [comment]
I don't think you need a PR to do this, I haven't tested this or thought too much about what the current or redoStack should be but something along these lines ought to work

```typescript
import { useState, useRef } from 'react';
import {createEmptyHistoryState, HistoryPlugin} from '@lexical/history';

export default function DiscardHistoryPlugin({ discard = false }: { discard: boolean }) {
  const [history, setHistory] = useState(createEmptyHistoryState);
  const historyRef = useRef(history);
  useEffect(() => {
    if (discard) {
      // save a copy of the history from before discard=true
      historyRef.current = {...history};
    } else {
      // restore the history redoStack & undoStack unless the initial state was discard=false
      setHistory((curHistory) => curHistory === historyRef.current ? curHistory : { ...historyRef.current, current: curHistory.current });
    }
  }, [discard]);
  return (<HistoryPlugin history={history} />);
}
```

**Comment by hanselninja** [comment]
> I don't think you need a PR to do this, I haven't tested this or thought too much about what the current or redoStack should be but something along these lines ought to work
> 
> ```ts
> import { useState, useRef } from 'react';
> import {createEmptyHistoryState, HistoryPlugin} from '@lexical/history';
> 
> export default function DiscardHistoryPlugin({ discard = false }: { discard: boolean }) {
>   const [history, setHistory] = useState(createEmptyHistoryState);
>   const historyRef = useRef(history);
>   useEffect(() => {
>     if (discard) {
>       // save a copy of the history from before discard=true
>       historyRef.current = {...history};
>     } else {
>       // restore the history redoStack & undoStack unless the initial state was discard=false
>       setHistory((curHistory) => curHistory === historyRef.current ? curHistory : { ...historyRef.current, current: curHistory.current });
>     }
>   }, [discard]);
>   return (<HistoryPlugin history={history} />);
> }
> ```

Honestly, not sure I follow you here. That snippet is not accomplishing the goal of this PR. Also, it should probably trigger an infinite loop once you pass `history` as a dependency.

**Comment by hanselabreu** [comment]
> I don't think you need a PR to do this, I haven't tested this or thought too much about what the current or redoStack should be but something along these lines ought to work
> 
> ```ts
> import { useState, useRef } from 'react';
> import {createEmptyHistoryState, HistoryPlugin} from '@lexical/history';
> 
> export default function DiscardHistoryPlugin({ discard = false }: { discard: boolean }) {
>   const [history, setHistory] = useState(createEmptyHistoryState);
>   const historyRef = useRef(history);
>   useEffect(() => {
>     if (discard) {
>       // save a copy of the history from before discard=true
>       historyRef.current = {...history};
>     } else {
>       // restore the history redoStack & undoStack unless the initial state was discard=false
>       setHistory((curHistory) => curHistory === historyRef.current ? curHistory : { ...historyRef.current, current: curHistory.current });
>     }
>   }, [discard]);
>   return (<HistoryPlugin history={history} />);
> }
> ```

Honestly, not sure I follow you here. That snippet is not accomplishing the goal of this PR. Also, it should probably trigger an infinite loop once you pass `history` as a dependency.

The problem is that there is no way to intercept editor changes when `discard = true` to prevent them from being pushed to the `undoStack`.

File: packages/lexical-history/src/index.ts [hunk 1]
```diff
@@ -244,7 +248,7 @@ function createMergeActionGetter(
 
     // If applying changes from history stack there's no need
     // to run history logic again, as history entries already calculated
-    if (tags.has('historic')) {
+    if (options.discardHistory || tags.has('historic')) {
       prevChangeType = OTHER;
       prevChangeTime = changeTime;
       return DISCARD_HISTORY_CANDIDATE;
```
Comment (hanselninja) [inline packages/lexical-history/src/index.ts:254]: Got the `discardHistory` name from here.

Comment (hanselabreu) [inline packages/lexical-history/src/index.ts:254]: Got the `discardHistory` name from here.

**Comment by etrepum** [comment]
Closing this PR due to staleness! If there are new updates, please reopen the PR.

**Pull request closed without merging**